scripts/all_the_kinds.py

```diff
@@ -0,0 +1,288 @@
+
+    if isinstance(expr, (int, float)):
+        start_zoom = expr
+
+    elif isinstance(expr, dict):
+        if expr.keys() == ['col']:
+            # if this is the special "zoom" column, then we know it's based on
+            # area, and theoretically could be zero???
+            if expr['col'] == 'zoom':
+                start_zoom = 0
+            else:
+                # this could come from anywhere, so we don't have a clue what
+                # it might be. therefore, return a value which means unknown.
+                start_zoom = None
+
+        elif expr.keys() == ['lookup']:
+            table = expr['lookup']['table']
+            start_zoom = min(row[0] for row in table)
+
+        elif expr.keys() == ['case']:
+            # take the min of all the branches
+            branches = []
+            for case in expr['case']:
+                if case.keys() == ['else']:
+                    branches.append(parse_start_zoom(case['else']))
+
+                else:
+                    assert sorted(case.keys()) == ['then', 'when']
+                    branches.append(parse_start_zoom(case['then']))
+
+            start_zoom = min(*branches)
+
+        elif expr.keys() == ['min']:
+            start_zoom = min(parse_start_zoom(item) for item in expr['min'])
+
+        elif expr.keys() == ['max']:
+            start_zoom = max(parse_start_zoom(item) for item in expr['max'])
+
+        elif expr.keys() == ['clamp']:
+            clamp = expr['clamp']
+            start_zoom = max(parse_start_zoom(clamp['min']),
+                             parse_start_zoom(clamp['value']))
+
+        elif expr.keys() == ['sum']:
+            start_zoom = sum(parse_start_zoom(item) for item in expr['sum'])
+
+        elif expr.keys() == ['mul']:
+            start_zoom = 1
+            for item in expr['mul']:
+                start_zoom *= parse_start_zoom(item)
+
+        elif expr.keys() == ['call']:
+            # we can perhaps figure this out for some functions and stash the
+            # results as constants?
+            func = expr['call']['func']
+            start_zoom = KNOWN_FUNCS.get(func)
+
+        else:
+            raise AssertionError("Unknown min zoom expression %r in %r"
+                                 % (expr.keys(), expr))
+
+    else:
+        raise AssertionError("Unknown min zoom expression type %s in %r"
+                             % (type(expr), expr))
+
+    return start_zoom
+
+
+# need this to pass into the CSVMatcher, which expects a Shapely geometry.
+FakeShape = namedtuple('FakeShape', 'type')
+
+
+def parse_item(layer_name, item, sort_rank):
+    kind = item['output'].get('kind')
+    if not kind:
+        return {}
+
+    if isinstance(kind, (str, unicode)):
+        all_kinds = [kind]
+
+    elif isinstance(kind, dict):
+        if kind.keys() == ['col']:
+            all_kinds = parse_filter_for_col(kind['col'], item['filter'])
+
+        elif kind.keys() == ['case']:
+            # TODO: figure this out
```

Comment:
this comment still apply?

Reply:
Nope! Well spotted, thanks. Removed in 4977583.